docs: add new `r/docs/...` examples

davd-gzl opened the pull request:
close: #3292

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: davd-gzl/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🔴 At least one of these user(s) reviewed the pull request: [aronpark1007 davd-gzl jefft0 notJoon omarsy MikaelVallenet] (with state "APPROVED")
    │       ├── 🟢 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🟢 Then
        └── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

jeronimoalbi commented on examples/gno.land/r/docs/factory/factory.gno:
```diff
@@ -0,0 +1,183 @@
+}
+
+func (f *FungibleToken) GetInfo() string {
+	return f.name + ": " + strconv.Itoa(int(f.amount)) + " units"
```
Nit

```suggestion
	return f.name + ": " + strconv.FormatUint(f.amount, 10) + " units"
```

davd-gzl replied:
Applied: `GetInfo` uses `strconv.FormatUint(f.amount, 10)`. https://github.com/gnolang/gno/pull/5016/commits/e61b07a3b9611f6d898707f93a6d39bfe16399c2 *(AI Agent)*

jeronimoalbi commented on examples/gno.land/r/docs/factory/factory.gno:
```diff
@@ -0,0 +1,183 @@
+		return "Insufficient balance"
+	}
+	f.amount -= amount
+	return "Transferred " + strconv.Itoa(int(amount)) + " " + f.name + " to " + to.String()
```
Nit

```suggestion
	return "Transferred " + strconv.FormatUint(amount, 10) + " " + f.name + " to " + to.String()
```

davd-gzl replied:
This line changed: `Transfer` now returns `error` instead of a formatted string, so the concat is gone (the same `FormatUint` fix landed on `GetInfo`). https://github.com/gnolang/gno/pull/5016/commits/9940005a84b7e805a4eaf1e55fb665880d7561b2 *(AI Agent)*

jeronimoalbi commented on examples/gno.land/r/docs/factory/factory.gno:
```diff
@@ -0,0 +1,183 @@
+type Token interface {
+	GetInfo() string
+	GetType() string
+	Transfer(to address, amount uint64) string
```
Expecting a panic, or returning error might make more sense:

```suggestion
	Transfer(to address, amount uint64) error
```

davd-gzl replied:
Done: `Transfer` now returns `error` (nil on success) on the interface and all three implementations, with the checks before any mutation. https://github.com/gnolang/gno/pull/5016/commits/9940005a84b7e805a4eaf1e55fb665880d7561b2 *(AI Agent)*

jeronimoalbi commented on examples/gno.land/r/docs/json/json.gno:
```diff
@@ -0,0 +1,191 @@
+	"gno.land/p/onbloc/json"
+)
+
+// User represents a simple user structure
+type User struct {
+	Name  string
+	Email string
+	Age   int
+}
+
```
I think is not being used:

```suggestion
```

davd-gzl replied:
Addressed: `CreateUserJSON` now builds the node from a `User` value, so the struct is used. https://github.com/gnolang/gno/pull/5016/commits/e61b07a3b9611f6d898707f93a6d39bfe16399c2 *(AI Agent)*

jeronimoalbi commented on examples/gno.land/r/docs/json/json.gno:
```diff
@@ -0,0 +1,191 @@
+	return lastCreatedJSON
+}
+
+// ParseJSON parses a JSON string and returns formatted output
+func ParseJSON(data string) (string, error) {
```
Nit

```suggestion
// PrettyPrintJSON parses a JSON string and returns formatted output
func PrettyPrintJSON(data string) (string, error) {
```

davd-gzl replied:
Done: renamed to `PrettyPrintJSON` with your comment. https://github.com/gnolang/gno/pull/5016/commits/9940005a84b7e805a4eaf1e55fb665880d7561b2 *(AI Agent)*

**Update:** `PrettyPrintJSON` was later removed entirely. It was not surfaced in `Render` and re-implemented `json.Marshal`/`json.Indent` by hand, which worked against the page's lesson, so there is no longer a function to rename. https://github.com/gnolang/gno/blob/877a57bd89e72d085bfa475ec57bf0a18b6b1bb3/examples/gno.land/r/docs/json/json.gno *(AI Agent)*

jeronimoalbi commented on examples/gno.land/r/docs/json/json.gno:
A test would provably be good to make sure pretty format works as expected

davd-gzl replied:
Added `json_test.gno` covering object, array, and invalid-input cases for `PrettyPrintJSON`. https://github.com/gnolang/gno/pull/5016/commits/9940005a84b7e805a4eaf1e55fb665880d7561b2 *(AI Agent)*

**Update:** the pretty-printer was later removed entirely, so `json_test.gno` (which only covered `PrettyPrintJSON`) was deleted along with it. Nothing left to test here. *(AI Agent)*

jeronimoalbi commented on examples/gno.land/r/docs/registry/registry.gno:
```diff
@@ -0,0 +1,274 @@
+
+// Register adds a new service to the registry
+func Register(_ realm, name, description, endpoint string) {
+	if name == "" {
+		panic("name cannot be empty")
+	}
```
Nit

```suggestion
	name = strings.TrimSpace(name)
	if name == "" {
		panic("name cannot be empty")
	}
```

It could be applied to the other string values, but if not having one case to set an example might be good enough.

davd-gzl replied:
Applied: `Register` does `name = strings.TrimSpace(name)` and panics when empty. https://github.com/gnolang/gno/pull/5016/commits/e61b07a3b9611f6d898707f93a6d39bfe16399c2 *(AI Agent)*

jeronimoalbi commented on examples/gno.land/r/docs/mistakes/mistakes.gno:
I'm thinking that other case to have in mind, which might lead to unwanted behavior is returning type references or `any`, which would potentially allow a caller to modify data though type methods for example. Is not a mistake per se, but it could be if there is no intention to allow modification though references.

davd-gzl replied:
Added a section 5, Returning References to Internal State, for this. Framed around Gno read-only taint: an external realm cannot write a returned field directly, but a mutating method on the value and same-realm code still can, so returning a copy is the safe choice. https://github.com/gnolang/gno/pull/5016/commits/9940005a84b7e805a4eaf1e55fb665880d7561b2 *(AI Agent)*

jeronimoalbi commented on examples/gno.land/r/docs/tdd/tdd.gno:
````diff
@@ -0,0 +1,144 @@
+
+## Quick Example
+
+First, write the test:
+
+` + "```go" + `
+// calculator_test.gno
+func TestAdd(t *testing.T) {
+    result := Add(nil, 2, 3)
+    if result != 5 {
+        t.Errorf("Add(nil, 2, 3) = %d; want 5", result)
+    }
+}
+` + "```" + `
````
Maybe, to make new users (readers) aware of the import too:

```suggestion
` + "```go" + `
import "testing"

// calculator_test.gno
func TestAdd(t *testing.T) {
    result := Add(nil, 2, 3)
    if result != 5 {
        t.Errorf("Add(nil, 2, 3) = %d; want 5", result)
    }
}
` + "```" + `
```

davd-gzl replied:
Done, applied your suggestion: the first `calculator_test.gno` snippet now shows `import "testing"`.

While here, I checked the other test snippets in r/docs for the same issue: the crossing doc had a test example calling `testing.SetRealm`/`testing.NewCodeRealm`/`testing.SetOriginCaller` without showing the `testing` import, so I added it there too. https://github.com/gnolang/gno/pull/5016/commits/38ac75e6795cff30756a911f30b69a1d6848a3ba *(AI Agent)*

jeronimoalbi commented on examples/gno.land/r/docs/soliditypatterns/banker/banker.gno:
```diff
@@ -0,0 +1,201 @@
+// This example demonstrates the Gno equivalent of Solidity's
+// payable functions and ETH transfers.
+
+var deposits avl.Tree // address -> int64
+
+func init() {
+	deposits = avl.Tree{}
+}
```
The init part is not needed in this case because of the way `avl.Tree` is implemented, no need for initialization:

```suggestion
var deposits avl.Tree // address -> int64
```

The `init` would make sense if:

```go
var deposits *avl.Tree // address -> int64

func init() {
	deposits = avl.NewTree()
}
```

Alternatively just:

```go
var deposits = avl.NewTree() // address -> int64
```

davd-gzl replied:
Applied: `var deposits avl.Tree` with no init. https://github.com/gnolang/gno/pull/5016/commits/e61b07a3b9611f6d898707f93a6d39bfe16399c2 *(AI Agent)*

jeronimoalbi commented on examples/gno.land/r/docs/crossing/crossing.gno:
```diff
@@ -0,0 +1,111 @@
+
+var owner address
+
+func init() {
+	owner = runtime.CurrentRealm().Address()
+}
```
```suggestion
func init() {
	// In this case there's no crossing, init function has no "realm" argument.
	// Function is called automatically only once on deployment, so "owner" is
	// initialized to the address of the realm being deployed.
	owner = runtime.CurrentRealm().Address()
}
```

davd-gzl replied:
Applied, using `unsafe.CurrentRealm().Address()` (the current non-crossing accessor) with your comment. https://github.com/gnolang/gno/pull/5016/commits/e61b07a3b9611f6d898707f93a6d39bfe16399c2 *(AI Agent)*

jeronimoalbi replied:
With interrealm V2 now I think this is also a possibility:
 ```suggestion
func init(cur realm) {
	owner = cur.Address()
}
```

davd-gzl replied:
Applied: `init` now takes `cur realm` and sets `owner = cur.Address()`, dropping the `unsafe.CurrentRealm()` call. Kept the comment version-neutral (no "V2" label) since this is just how init works now. https://github.com/gnolang/gno/pull/5016/commits/160789c930ddad7a6a3ac925452584cbf070c36c

*(AI Agent)*

jeronimoalbi commented on examples/gno.land/r/docs/crossing/crossing.gno:
```diff
@@ -0,0 +1,111 @@
+
+// InternalHelper is a helper function (no realm parameter)
+// It can still be called from other realms but is typically
+// used internally within this realm
+func InternalHelper() string {
+	return "I'm a helper function"
+}
```
```suggestion
func InternalHelper() string {
	// In this case there's no crossing, InternalHelper function has no "realm" argument.
	// Caller is initialized to the address of the caller and NOT to the address
	// of the "gno.land/r/docs/crossing" realm, because there is no crossing.
	caller := runtime.CurrentRealm().Address()
	return "I'm a helper function called by: " + caller.String()
}
```

davd-gzl replied:
Applied here too, with `unsafe.CurrentRealm().Address()`. https://github.com/gnolang/gno/pull/5016/commits/e61b07a3b9611f6d898707f93a6d39bfe16399c2 *(AI Agent)*

jeronimoalbi replied:
With interrealm V2 now something like this is also a possibility:

```suggestion
func InternalHelper(_ int, rlm realm) string {
	if !rlm.IsCurrent() {
		panic("rlm does not match the current crossing frame")
	}

	caller := rlm.Address()
	return "I'm a helper function called by: " + caller.String()
}
```

davd-gzl replied:
That's right, thank you. I've added another helper to showcase that precise use case: https://github.com/gnolang/gno/pull/5016/commits/5e7f8e2e423c3870d7d09f4219b8f6c50a59b2e7

davd-gzl committed:
docs: add new r/docs/... examples

Add example realms demonstrating Gno patterns and concepts:
- adder, buttons, complexargs, composition, crossing
- factory, json, mistakes, registry
- soliditypatterns (banker, ownable, reentrancy)
- tdd, userprofile

Address review feedback from jeronimoalbi:
- registry: add TrimSpace for name validation
- crossing: add explanatory comments to init() and InternalHelper()
- tdd: add missing import in test example
- json: clarify ParseJSON doc comment

Migrate examples to chain/runtime/unsafe and gno.land/p/nt/avl/v0.

davd-gzl committed:
fix(examples/docs): fix crossing-call form in tdd test, address review

tdd_test.gno used bare `cross`; current gno only accepts `cur` or `cross(rlm)`.
Genesis loads example test files and typechecks them, so this broke node
startup and the lint/test CI jobs. Switch to `TestX(cur realm, t *testing.T)`
with `cross(cur)`.

Also fix outdated tdd/crossing doc snippets (cross(cur), revive,
testutils.TestAddress, NewCodeRealm), rename json ParseJSON to PrettyPrintJSON
and add a test, make factory Transfer return error, and add a "returning
internal references" section to mistakes.

davd-gzl committed:
Merge origin/master into david/rdocs-additions

davd-gzl committed:
docs(rdocs): show "import \"testing\"" in crossing test example

davd-gzl committed:
Merge branch 'master' into david/rdocs-additions

davd-gzl committed:
docs(rdocs): idiomatic cur.Previous() caller auth + clearer crossing doc

davd-gzl committed:
docs(rdocs): add package summaries, simplify json realm

davd-gzl committed:
docs(rdocs): fix doc-example bugs across r/docs realms

davd-gzl committed:
docs(rdocs/json): clarify lastCreatedJSON comment

davd-gzl committed:
Merge branch 'master' into david/rdocs-additions

davd-gzl committed:
docs(rdocs/crossing): set owner via cur realm in init

davd-gzl committed:
docs(rdocs/crossing): add InternalHelperWithRealm example (realm param, IsCurrent)

jeronimoalbi commented on examples/quarantined/gno.land/r/docs/soliditypatterns/banker/banker.gno:
```diff
@@ -0,0 +1,196 @@
+	"strconv"
+
+	"gno.land/p/nt/avl/v0"
+)
+
+// This example demonstrates the Gno equivalent of Solidity's
+// payable functions and ETH transfers.
+
+var deposits avl.Tree // address -> int64
+
+// Deposit accepts coins and tracks the sender's balance
+// In Solidity, this would be a payable function
+func Deposit(cur realm) {
+	caller := cur.Previous().Address()
+	sent := unsafe.OriginSend()
+
+	// Check that coins were sent
+	amount := sent.AmountOf("ugnot")
+	if amount <= 0 {
+		panic("must send some ugnot")
+	}
+
+	// Get current balance
+	currentBalance := int64(0)
+	if val, exists := deposits.Get(caller.String()); exists {
+		currentBalance = val.(int64)
+	}
+
+	// Update balance
+	deposits.Set(caller.String(), currentBalance+amount)
+}
```
It seems that this case needs to assert that is a user call `cur.Previous().IsUserCall()`:

```suggestion
func Deposit(cur realm) {
	if !cur.Previous().IsUserCall() {
		panic("invalid non-user call")
	}

	caller := cur.Previous().Address()
	sent := unsafe.OriginSend()

	// Check that coins were sent
	amount := sent.AmountOf("ugnot")
	if amount <= 0 {
		panic("must send some ugnot")
	}

	// Get current balance
	currentBalance := int64(0)
	if val, exists := deposits.Get(caller.String()); exists {
		currentBalance = val.(int64)
	}

	// Update balance
	deposits.Set(caller.String(), currentBalance+amount)
}
```

jeronimoalbi replied:
Same for `Withdraw()`

davd-gzl replied:
Guard is in (your accepted suggestion); added a short comment above the assertion explaining it in [`6527ddc134137ee84a4427ca63d418199006342d`](https://github.com/gnolang/gno/pull/5016/commits/6527ddc134137ee84a4427ca63d418199006342d).

*(AI Agent)*

davd-gzl replied:
Done: added the same `cur.Previous().IsUserCall()` guard to `Withdraw()` in [`6527ddc134137ee84a4427ca63d418199006342d`](https://github.com/gnolang/gno/pull/5016/commits/6527ddc134137ee84a4427ca63d418199006342d).

*(AI Agent)*

jeronimoalbi commented on examples/quarantined/gno.land/r/docs/soliditypatterns/reentrancy/reentrancy.gno:
Consider adding `cur.Previous().IsUserCall()` assertion

davd-gzl replied:
Done: added the guard to the `Withdraw` snippets in [`6527ddc134137ee84a4427ca63d418199006342d`](https://github.com/gnolang/gno/pull/5016/commits/6527ddc134137ee84a4427ca63d418199006342d).

*(AI Agent)*

davd-gzl committed:
Update examples/gno.land/r/docs/soliditypatterns/banker/banker.gno

Co-authored-by: Jerónimo Albi <894299+jeronimoalbi@users.noreply.github.com>

davd-gzl committed:
docs(rdocs/soliditypatterns): guard Withdraw with IsUserCall

Address review: assert cur.Previous().IsUserCall() on Withdraw (banker
realm and reentrancy snippets) alongside the accepted Deposit guard, and
add a short comment explaining each assertion.

davd-gzl committed:
Merge branch 'master' into david/rdocs-additions

alexiscolin committed:
Merge branch 'master' into david/rdocs-additions

davd-gzl committed:
Merge branch 'master' into david/rdocs-additions

davd-gzl committed:
docs(rdocs): guard cur.Previous() caller reads with cur.IsCurrent()

Add `if !cur.IsCurrent() { panic("spoofed realm") }` before every
cur.Previous() caller-identity read across the new r/docs realms
(composition, factory, registry, crossing, banker, reentrancy,
mistakes), matching #5871. A non-crossing call could otherwise pass a
non-current realm handle so cur.Previous() names a forged caller.

davd-gzl committed:
docs(rdocs): drop json example

Remove the gno.land/r/docs/json realm and its home index link.

davd-gzl committed:
docs(rdocs): move new examples into examples/quarantined

r/docs lives under examples/quarantined since #5726 (non-test-13 set).
Place the new realms alongside their siblings and the home/soliditypatterns
indexes so #5016 stays consistent with master.

davd-gzl committed:
docs(rdocs/userprofile): use cross(cur), not bare cross

gfanton commented on examples/quarantined/gno.land/r/docs/userprofile/userprofile.gno:
```diff
@@ -0,0 +1,159 @@
+addr := data.Addr()
+
+// Resolve an address to user data
+data, _ = users.ResolveAddress(someAddress)
```
This will not compile, as ResolveAddress return only one parameter.

davd-gzl replied:
Fix: https://github.com/gnolang/gno/pull/5016/commits/f30274440a994e789bef94fb20a23e62c17c6e50

davd-gzl committed:
docs(rdocs): explain spoofed-realm guard in snippets; fix banker reentrancy note

Snippet guards now carry a one-line note on what cur.IsCurrent() rejects.
Banker's withdraw snippet no longer claims state-before-send 'prevents
reentrancy' (Gno has none); reframed as checks-effects-interactions.

davd-gzl committed:
docs(rdocs/userprofile): ResolveAddress returns one value, not two

gfanton commented on examples/quarantined/gno.land/r/docs/registry/registry.gno:
```diff
@@ -0,0 +1,291 @@
+
+| Property | Value |
+|----------|-------|
+| Status | ` + status + ` |
+| Description | ` + service.Description + ` |
+| Endpoint | ` + service.Endpoint + ` |
+| Owner | ` + service.Owner.String() + ` |
```
Usanetize user input, use `markdown/sanitize`

davd-gzl replied:
Fix: https://github.com/gnolang/gno/pull/5016/commits/b70328606242fa2e93c21d66f977dba00d285301

davd-gzl committed:
docs(rdocs/factory): CreateToken snippet param is TokenType, not int

gfanton commented on examples/quarantined/gno.land/r/docs/registry/registry.gno:
```diff
@@ -0,0 +1,291 @@
+		return false
+	})
+	return result
+}
```
Both function return internal pointer, while its harmless in pratice,  it contradict `mistakes.gno`.  so either copy / or explain why its acceptable in this case.

davd-gzl replied:
Fix: https://github.com/gnolang/gno/pull/5016/commits/d1f7df06d9aa303408a4499add311f686327c078

davd-gzl committed:
docs(rdocs): sanitize user input in registry/userprofile Render

Echoed user-controlled strings (service name/description/endpoint, lookup
paths, profile fields) through p/nt/markdown/sanitize/v0 before they land in
Render markdown: InlineText for inline text, TableCell for table cells.
profileCell keeps the _not set_ placeholder for empty profile fields.

davd-gzl committed:
docs(rdocs/registry): return service copies, not internal pointers

Apply gno-security-guide.md (5.1 / 8 checklist): Lookup and GetAllServices
returned the avl-stored *Service, aliasing internal mutable state, which
contradicts the realm's own /r/docs/mistakes lesson. Return value copies
instead; update the Code Pattern snippet to match. Addresses gfanton review.

davd-gzl committed:
docs(rdocs/mistakes): drop third-party uint256 library recommendation

davd-gzl committed:
docs(rdocs): panic("invalid realm") instead of "spoofed realm"

Rename the cur.IsCurrent() guard panic across the new r/docs realms and
align the guard comments from "spoofed" to "non-current realm handle".

davd-gzl committed:
docs(rdocs): drop duplicated IsCurrent-guard comments

The same guard comment was pasted at ~9 sites. Keep the explanation in
crossing (the realm about it) and the mistakes prose; leave the bare
guard everywhere else.

davd-gzl committed:
docs(rdocs/registry): state the copy rationale once

Lookup, GetAllServices and the snippet all repeated the same
copy-not-pointer / see-mistakes rationale. Keep it on Lookup; the others
just say they return copies.

davd-gzl committed:
fix(rdocs/banker): reject non-positive withdraw amount

A negative amount passed the currentBalance < amount check, inflated the
caller's balance (balance - negative) and sent negative coins. Guard with
amount <= 0, matching Deposit.

davd-gzl committed:
Merge branch 'master' into david/rdocs-additions

davd-gzl committed:
fix(rdocs): adapt registry and banker to single-value avl.Tree.Get

avl.Tree.Get now returns a single any value, nil when the key is absent,
instead of (any, bool). The registry and banker examples still used the
two-value form, so the genesis type check rejected them and the e2e node
panicked at InitChain.

Fold each call into a comma-ok type assertion. An absent key yields a nil
interface, whose assertion reports false, preserving the previous
not-found behavior. Also update the Lookup sample embedded in the registry
Render output, which taught the stale API.

davd-gzl committed:
style(rdocs): use the nil-check form for avl.Tree.Get lookups

The previous commit asserted directly on the Get result, the pattern the
avl doc comment suggests. The rest of the tree does not use it: 129 call
sites check the returned any against nil and assert separately, and no
call site outside a doc comment uses the assertion form.

Match the surrounding code, since these realms exist to teach the idiom.
GetBalance now mirrors grc20's BalanceOf. Behavior is unchanged: neither
realm stores a nil value, so nothing distinguishes the two forms.

davd-gzl committed:
style(rdocs): guard avl lookups with Has

Get returns nil for an absent key and for a key stored with a nil value,
so a nil check cannot tell them apart. The avl doc comment says to use Has
for existence. Guard with Has, then assert on Get.

This also matches Register and Delete in registry, which already guard
with Has. Behavior is unchanged.

davd-gzl committed:
Merge branch 'master' into david/rdocs-additions

davd-gzl committed:
style(rdocs): read avl values with a single Get

Guarding with Has and then calling Get walks the tree twice for one
lookup. Where the value is needed, call Get once and check the result
against nil. Register and Delete keep Has, since they only test for
existence and never read the value.

Reverts the lookup sites changed in 145831511. Behavior is unchanged:
neither realm stores a nil value.

davd-gzl commented:
Fixed CI: https://github.com/gnolang/gno/pull/5016/changes/52b21ad0bd2bd0555bcf73957510cd92a4cc7309..07362d202994c57195b40528c1a72863e44e7824

thehowl committed:
fix(rdocs): sanitize registry link URL, correct reentrancy snippet

- registry: pass the service link destination through sanitize.URL so a user-controlled service name can't break out of the markdown link
- reentrancy: the "why Gno is safe" snippet now debits amount instead of zeroing the whole balance
- factory: drop a vestigial trailing-newline append in Render
- trim accidental trailing whitespace (crossing.gno diagram left intact)

thehowl committed:
Merge branch 'master' into david/rdocs-additions
